zebra/kernel_netlink.c

```diff
@@ -697,6 +697,204 @@ static void netlink_parse_extended_ack(struct nlmsghdr *h)
 	}
 }
 
+/*
+ * netlink_send_msg - send a netlink message of a certain size.
+ *
+ * Returns -1 on error. Otherwise, it returns the number of bytes sent.
+ */
+static int netlink_send_msg(const struct nlsock *nl, void *buf, size_t buflen)
+{
+	struct sockaddr_nl snl;
+	struct iovec iov;
+	struct msghdr msg;
+	int status, save_errno = 0;
+
+	memset(&snl, 0, sizeof(snl));
+	memset(&iov, 0, sizeof(iov));
+	memset(&msg, 0, sizeof(msg));
+
+	iov.iov_base = buf;
+	iov.iov_len = buflen;
+	msg.msg_name = (void *)&snl;
+	msg.msg_namelen = sizeof(snl);
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	snl.nl_family = AF_NETLINK;
+
+	/* Send message to netlink interface. */
+	frr_with_privs(&zserv_privs) {
+		status = sendmsg(nl->sock, &msg, 0);
+		save_errno = errno;
+	}
+
+	if (IS_ZEBRA_DEBUG_KERNEL_MSGDUMP_SEND) {
+		zlog_debug("%s: >> netlink message dump [sent]", __func__);
+		zlog_hexdump(buf, buflen);
+	}
+
+	if (status < 0) {
+		flog_err_sys(EC_LIB_SOCKET, "%s error: %s", __func__,
+			     safe_strerror(save_errno));
+		return -1;
+	}
+
+	return status;
+}
+
+/*
+ * netlink_recv_msg - receive a netlink message.
+ *
+ * Returns -1 on error, 0 if read would block or the number of bytes received.
+ */
+static int netlink_recv_msg(const struct nlsock *nl, struct msghdr msg,
+			    void *buf, size_t buflen)
+{
+	struct iovec iov;
+	int status;
+
+	iov.iov_base = buf;
+	iov.iov_len = buflen;
+	msg.msg_iov = &iov;
+	msg.msg_iovlen = 1;
+
+	do {
+#if defined(HANDLE_NETLINK_FUZZING)
+		/* Check if reading and filename is set */
+		if (netlink_read && '\0' != netlink_fuzz_file[0]) {
+			zlog_debug("Reading netlink fuzz file");
+			status = netlink_read_file(buf, netlink_fuzz_file);
+			((struct sockaddr_nl *)msg.msg_name)->nl_pid = 0;
+		} else {
+			status = recvmsg(nl->sock, &msg, 0);
+		}
+#else
+		status = recvmsg(nl->sock, &msg, 0);
+#endif /* HANDLE_NETLINK_FUZZING */
+	} while (status < 0 && errno == EINTR);
```

Comment:
**Non blocking comment**

Small nit:
`recvmsg` and other syscalls only return `-1` on error, if it is returning something else I think a lot other places will break.

I feel bad about commenting this, because the new code really looks better and this is just a style thing.

Feel free to ignore this.

Reply:
I changed this too.